walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,70 @@
 
 package com.google.sps;
 
+import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.HashSet;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    // throw new UnsupportedOperationException("TODO: Implement this method.");
+    final long MINUTES_IN_A_DAY = 1440;
+    final int START_OF_DAY = TimeRange.START_OF_DAY;
+    final int END_OF_DAY = TimeRange.END_OF_DAY;
+    int endOfTimeRange = 0;
+    TimeRange previousTimeRange = TimeRange.fromStartEnd(0,0,true);
+    Collection<TimeRange> unavailableTimeRange = new ArrayList<>();
+    Collection<TimeRange> unavailableNoOverlap = new ArrayList<>();
+    Collection<TimeRange> returnCollection = new ArrayList<>();
+
+    long durationOfRequest = request.getDuration();
+    Collection<String> attendeesOfRequest = request.getAttendees();
+
+    if(attendeesOfRequest.isEmpty()){
+      return Arrays.asList(TimeRange.WHOLE_DAY);
+    }else if(durationOfRequest > MINUTES_IN_A_DAY){
+      return Arrays.asList();
+    }else{
+      for(String attendee : attendeesOfRequest){
+        for(Event event : events){
+          if(event.getAttendees().contains(attendee)){
+            unavailableTimeRange.add(event.getWhen());
+          }
+        }
+      }
+      for(TimeRange timerange : unavailableTimeRange){
+        if(previousTimeRange.contains(timerange)){
+        }else if(previousTimeRange.overlaps(timerange)){
+          TimeRange newTimeRange = TimeRange.fromStartEnd(previousTimeRange.start(),timerange.end(),false);
+          unavailableNoOverlap.add(newTimeRange);
+          unavailableNoOverlap.remove(previousTimeRange);
+          previousTimeRange = newTimeRange;
+        }else{
+          unavailableNoOverlap.add(timerange);
+          previousTimeRange = timerange;
+        }
+      }
+      for(TimeRange timerange : unavailableNoOverlap){
+        if(endOfTimeRange == 0){
+          TimeRange trval = TimeRange.fromStartEnd(START_OF_DAY,timerange.start(),false);
+          if(trval.duration() != 0 && trval.duration() >= durationOfRequest){
+            returnCollection.add(trval);
+          }
```

Comment:
You can move these lines before the `for(TimeRange timerange : unavailableNoOverlap)`, and then you don't have to have an if/else statement inside

Reply:
I cant shift it to the top before the for loop as i need the first value's start timing which is the timerange.start().

Reply:
Actually it looks like you're doing the same thing in the `if` and `else` branches, since START_OF_DAY  = 0 which would be the same as endOfTimeRange for the first iteration. Simplify this by just doing:

```
for(TimeRange timerange : unavailableNoOverlap){
  TimeRange trval = TimeRange.fromStartEnd(endOfTimeRange,timerange.start(),false);
   if(trval.duration() != 0 && trval.duration() >= durationOfRequest){
      returnCollection.add(trval);
   }
  endOfTimeRange = timerange.end();
}
```